### ISSUE TITLE
Query: ArgumentOutOfRangeException when using Include and OrderBy by with casting

### ISSUE THREAD

**Issue by tomaszmadeyski** [aspnet/EntityFrameworkCore#10635]
This is somehow related with #9758 but I believe the root of issue might be different so I'm adding a new issue.

When I'm trying to do `OrderBy` which follows `Include` I'm getting `ArgumentOutOfRangeException`. What is important that it happens when casting is involved **and** with `Include`.

`var notOk = ctx.Parent.Include(p => p.Children).OrderBy(p => ((IEntity) p).Id).FirstOrDefault();`

Casting is in place because I'm calling `OrderBy` in generic method. 

```
Exception message:
ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
Stack trace:

System.ThrowHelper.ThrowArgumentOutOfRange_IndexException()
System.Collections.Generic.List.get_Item(int index)
Microsoft.EntityFrameworkCore.Query.Expressions.SelectExpression.BindSubqueryProjectionIndex(int projectionIndex, IQuerySource querySource)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.SqlTranslatingExpressionVisitor.Visit(Expression expression)
Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitOrderByClause(OrderByClause orderByClause, QueryModel queryModel, int index)
Remotion.Linq.Clauses.OrderByClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, int index)
Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection<IBodyClause> bodyClauses, QueryModel queryModel)
Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitSubQuery(SubQueryExpression expression)
Remotion.Linq.Clauses.Expressions.SubQueryExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.ExpressionVisitorBase.VisitLambda<T>(Expression<T> node)
System.Linq.Expressions.Expression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitMethodCall(MethodCallExpression node)
System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
System.Dynamic.Utils.ExpressionVisitorUtils.VisitBlockExpressions(ExpressionVisitor visitor, BlockExpression block)
System.Linq.Expressions.ExpressionVisitor.VisitBlock(BlockExpression node)
System.Linq.Expressions.BlockExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.ExpressionVisitorBase.VisitLambda<T>(Expression<T> node)
System.Linq.Expressions.Expression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
System.Dynamic.Utils.ExpressionVisitorUtils.VisitArguments(ExpressionVisitor visitor, IArgumentProvider nodes)
System.Linq.Expressions.ExpressionVisitor.VisitMethodCall(MethodCallExpression node)
Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.RelationalEntityQueryableExpressionVisitor.VisitMethodCall(MethodCallExpression node)
System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.ReplaceClauseReferences(Expression expression, IQuerySource querySource, bool inProjection)
Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitSelectClause(SelectClause selectClause, QueryModel queryModel)
Remotion.Linq.Clauses.SelectClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel)
Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor<TResult>(QueryModel queryModel)
Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery<TResult>(QueryModel queryModel)
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore<TResult>(Expression query, INodeTypeProvider nodeTypeProvider, IDatabase database, IDiagnosticsLogger<Query> logger, Type contextType)
Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler+<>c__DisplayClass15_0.<Execute>b__0()
Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore<TFunc>(object cacheKey, Func<Func<QueryContext, TFunc>> compiler)
Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery<TResult>(object cacheKey, Func<Func<QueryContext, TResult>> compiler)
Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute<TResult>(Expression query)
Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute<TResult>(Expression expression)
Remotion.Linq.QueryableBase.GetEnumerator()
System.Collections.Generic.List.AddEnumerable(IEnumerable<T> enumerable)
System.Collections.Generic.List..ctor(IEnumerable<T> collection)
System.Linq.Enumerable.ToList<TSource>(IEnumerable<TSource> source)
```

### Steps to reproduce
It's slight modification of @sveinungh example from #9758. No data in database is required 

```c#
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;

namespace EFDebug
{
    class Program
    {
        static void Main(string[] args)
        {
            var loggerFactory = new LoggerFactory();
            loggerFactory.AddDebug();
            var connection = @"";
            var ctx = new MyContext(new DbContextOptionsBuilder().UseSqlServer(connection).UseLoggerFactory(loggerFactory).Options);
            //this works ok
            var ok = ctx.Parent.Include(p => p.Children).OrderBy(p => p.Id).ToList();
            //this fails
            var notOk = ctx.Parent.Include(p => p.Children).OrderBy(p => ((IEntity) p).Id).ToList()
        }
    }
    public class Parent : IEntity
    {
        public int Id { get; set; }
        public string Name { get; set; }        

        public virtual ICollection<Child> Children { get; set; }


    }

    public class Child : IEntity
    {
        public int Id { get; set; }
        public string Name { get; set; }


        public int ParentId { get; set; }       
    }

    public interface IEntity
    {
        int Id { get; set; }
    }
    public class MyContext : DbContext
    {
        public MyContext(DbContextOptions options)
            : base(options)
        {

        }
        public DbSet<Parent> Parent { get; set; }
    }
}
/*
create table Parent
(
	Id int identity primary key,
	Name varchar(255)	
)

create table Child
(
	Id int identity primary key,
	Name varchar(255),
	ParentId int references parent(id)
)
*/
```

### Further technical details
EF Core version: 2.0.1
Database Provider:  Microsoft.EntityFrameworkCore.SqlServer (this does not happen in inMemory database)
Operating system: Windows 10
IDE: Rider 2017.3

**Comment by maumar** [comment]
Problem here is that include introduces orderby p.Id itself but when we are rewriting tree for include we don't recognize that p.Id and ((IEntity)p).Id is refers to the same thing. We assume those will be two different columns in the generated SQL, but in fact during sql generation we recognize them as being the same and only one column is being produced. We need to improve logic in CollectionQueryModelRewritingExpressionVisitor.BuildParentOrderings so that we recognize both orderbys to be the same earlier.

**Comment by maumar** [comment]
fixed in 01372d5a45397216677f1fdd062e6039730b672c
